pkg/k8s/watchers/cilium_network_policy.go

```diff
@@ -191,6 +277,98 @@ func (k *K8sWatcher) onDeleteCNP(
 	return err
 }
 
+func (k *K8sWatcher) onUpsertCIDRGroup(
+	cidrGroup *cilium_v2_alpha1.CiliumCIDRGroup,
+	cidrGroupCache map[string]*cilium_v2_alpha1.CiliumCIDRGroup,
+	cnpCache map[resource.Key]*types.SlimCNP,
+	cs client.Clientset,
+) error {
+	var errs []error
+
+	initialRecvTime := time.Now()
+
+	oldCidrGroup, ok := cidrGroupCache[cidrGroup.Name]
+	if ok && oldCidrGroup.Spec.DeepEqual(&cidrGroup.Spec) {
+		return nil
+	}
+
+	cidrGroupCpy := cidrGroup.DeepCopy()
+	cidrGroupCache[cidrGroup.Name] = cidrGroupCpy
+
+	for key, cnp := range cnpCache {
+		if !hasCIDRGroupRef(cnp, cidrGroup.Name) || cnp.RequiresDerivative() {
+			continue
+		}
+
+		// We need to deepcopy this structure because we are writing
+		// fields.
+		// See https://github.com/cilium/cilium/blob/27fee207f5422c95479422162e9ea0d2f2b6c770/pkg/policy/api/ingress.go#L112-L134
+		cnpCpy := cnp.DeepCopy()
+
+		translatedCNP := resolveCIDRGroupRef(cnp, cidrGroupCache)
```

Comment:
Again, this should be using the `cnpCpy`

```suggestion
		translatedCNP := resolveCIDRGroupRef(cnpCpy, cidrGroupCache)
```